release_notes/fetch.py

```diff
@@ -0,0 +1,217 @@
+import logging
+import typing
+
+import gci.componentmodel
+import git
+import github3.repos
+import semver
+
+import cnudie.retrieve
+import cnudie.util
+import version
+from github.release_notes.util import git_helper_from_github_access, github_helper_from_github_access
+from release_notes import association, models
+
+logger = logging.getLogger(__name__)
+
+
+def _find_previous_version(available_versions: list[semver.VersionInfo],
+                           current_version: semver.VersionInfo) -> typing.Optional[semver.VersionInfo]:
+    # find version before the requested version and sort by semver
+    return max((z for z in sorted(available_versions) if z < current_version), default=None)
+
+
+def _iter_tags(repo: git.Repo, newest_tag: git.TagReference, oldest_tag: git.TagReference) -> tuple[git.Commit]:
+    """ Iterates between two tags, even if they aren't linear.
+    """
+    if not (merge_commit_list := repo.merge_base(newest_tag, oldest_tag)):
+        raise RuntimeError("cannot find merge base")
+    # even though it's a list, it contains max. 1 commit since the --all flag is not passed.
+    merge_commit: git.Commit = merge_commit_list.pop()
+
+    # if the previous tag is after the merge-base, return all commits between merge-base and current tag
+    if merge_commit.authored_date < oldest_tag.commit.authored_date:
+        return tuple(repo.iter_commits(f"{newest_tag.commit.hexsha}...{merge_commit.hexsha}"))
+
+    # otherwise the tags should be linear
+    return tuple(repo.iter_commits(f"{newest_tag.commit.hexsha}...{oldest_tag.commit.hexsha}"))
+
+
+class ReleaseNotesComponentAccess:
+    def __init__(self,
+                 component: gci.componentmodel.Component,
+                 repo_path: str,
+                 current_version: typing.Optional[semver.VersionInfo] = None):
```

Comment:
also (I write this after having read through this class's methods) the name is completely misleading. I get it that it may be useful to have a struct to pass-around all those pieces of data (totally fine). however, I think those methods should be  turned into functions. use a module to group them, maybe. preferably pass-in the actually-used objects. this may look like some tedious overhead considering the amount of parameters, but I think this is just a couple of line of extra code, which will make the code a lot easier to understand considering its side-effects and semantics.

consider factoring out some of the special-cases of those "list"-methods and rather handle those in fewer functions (maybe just one)